core/src/main/scala/kafka/controller/ReplicaStateMachineV2.scala

```diff
@@ -0,0 +1,318 @@
+    }
+    results
+  }
+
+  private def doRemoveReplicasFromIsr(replicaId: Int, partitions: Seq[TopicAndPartition]) = {
+    val (currentLeaderAndIsrs, candidateLeaderAndIsrs, partitionsWithNoLeaderAndIsrInZk, failedStateReads) = getTopicPartitionStatesFromZk(replicaId, partitions)
+    val partitionsWhoseIsrExcludesReplica = currentLeaderAndIsrs -- candidateLeaderAndIsrs.keys
+    val (partitionsWithEmptyIsr, partitionsWithNonEmptyIsr) = candidateLeaderAndIsrs.partition { case (partition, leaderAndIsr) => leaderAndIsr.isr.isEmpty }
+    val (adjustedLeaderAndIsrsForPartitionsWithEmptyIsr, failedLogConfigReads) = leaderAndIsrBasedOnLogConfigs(currentLeaderAndIsrs, partitionsWithEmptyIsr)
+    val adjustedLeaderAndIsrs = adjustedLeaderAndIsrsForPartitionsWithEmptyIsr ++ partitionsWithNonEmptyIsr
+    val (successfulUpdates, updatesToRetry) = updateLeaderAndIsrToZk(adjustedLeaderAndIsrs)
+    partitionsWithNoLeaderAndIsrInZk.foreach { partition =>
+      if (!controller.topicDeletionManager.isPartitionToBeDeleted(partition)) {
+        val replica = PartitionAndReplica(partition.topic, partition.partition, replicaId)
+        val exception = new StateChangeFailedException(s"Failed to change state of replica $replicaId for partition $partition since the leader and isr path in zookeeper is empty")
+        logFailedStateChange(replica, replicaState(replica), OfflineReplica, exception)
+      }
+    }
+    (failedStateReads ++ failedLogConfigReads).foreach { case (partition, e) =>
+      val replica = PartitionAndReplica(partition.topic, partition.partition, replicaId)
+      logFailedStateChange(replica, replicaState(replica), OfflineReplica, e)
+    }
+    val leaderIsrAndControllerEpochs = (partitionsWhoseIsrExcludesReplica ++ successfulUpdates).map { case (partition, leaderAndIsr) =>
+      val leaderIsrAndControllerEpoch = LeaderIsrAndControllerEpoch(leaderAndIsr, controller.epoch)
+      controllerContext.partitionLeadershipInfo.put(partition, leaderIsrAndControllerEpoch)
+      partition -> leaderIsrAndControllerEpoch
+    }
+    (leaderIsrAndControllerEpochs, updatesToRetry)
+  }
+
+  // replicas not in isr are implied as partitions in currentLeaderAndIsrs that are not in candidateLeaderAndIsrs
+  private def getTopicPartitionStatesFromZk(replicaId: Int, partitions: Seq[TopicAndPartition]):
+  (Map[TopicAndPartition, LeaderAndIsr],
+    Map[TopicAndPartition, LeaderAndIsr],
+    Seq[TopicAndPartition],
+    Map[TopicAndPartition, Exception]) = {
+    val currentLeaderAndIsrs = mutable.Map.empty[TopicAndPartition, LeaderAndIsr]
+    val candidateLeaderAndIsrs = mutable.Map.empty[TopicAndPartition, LeaderAndIsr]
+    val partitionsWithNoLeaderAndIsrInZk = mutable.Buffer.empty[TopicAndPartition]
+    val failed = mutable.Map.empty[TopicAndPartition, Exception]
+    val getDataResponses = try {
+      KafkaControllerZkUtils.getTopicPartitionStates(zookeeperClient, partitions.zip(partitions).toMap)
+    } catch {
+      case e: Exception =>
```

Comment:
It's possible since it internally calls ZookeeperClient.waitUntilConnected, which can throw exceptions (ZookeeperClientAuthFailedException and ZookeeperClientExpiredException).

Reply:
Hmm, for both ZookeeperClientAuthFailedException and ZookeeperClientExpiredException, we probably don't want to retry forever in removeReplicasFromIsr(). It seems that we should just log an error and move on.

Reply:
It's not terribly clear but that's actually the behavior in the PR.